File: india_compliance/gst_india/report/gst_sales_register_beta/gst_sales_register_beta.js
// Copyright (c) 2024, Resilient Tech and contributors
// For license information, please see license.txt
const INVOICE_TYPE = {
    "B2B, SEZ, DE": [
        "B2B Regular",
        "B2B Reverse Charge",
        "SEZWP",
        "SEZWOP",
        "Deemed Exports",
    ],
    "B2C (Large)": ["B2C (Large)"],
    "Exports": ["EXPWP", "EXPWOP"],
    "B2C (Others)": ["B2C (Others)"],
    "Nil-Rated, Exempted, Non-GST": ["Nil-Rated", "Exempted", "Non-GST"],
    "Credit/Debit Notes (Registered)": ["CDNR"],
    "Credit/Debit Notes (Unregistered)": ["CDNUR"],
};

frappe.query_reports["GST Sales Register Beta"] = {
    onload: set_sub_category_options,

    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            on_change: report => {
                report.set_filter_value({
                    company_gstin: "",
                });
            },
            get_query: function () {
                return {
                    filters: {
                        country: "India",
                    },
                };
            },
            reqd: 1,
        },
        {
            fieldname: "company_gstin",
            label: __("Company GSTIN"),
            fieldtype: "Autocomplete",
            get_query() {
                const company = frappe.query_report.get_filter_value("company");
                return india_compliance.get_gstin_query(company);
            },
        },
        {
            fieldname: "date_range",
            label: __("Date Range"),
            fieldtype: "DateRange",
            default: [
                india_compliance.last_month_start(),
                india_compliance.last_month_end(),
            ],
            width: "80",
        },
        {
            fieldtype: "Select",
            fieldname: "summary_by",
            label: __("Summary By"),
            options: "Overview\nSummary by HSN\nSummary by Item",
            default: "Summary by Item",
        },
        {
            fieldtype: "Autocomplete",
            fieldname: "invoice_category",
            label: __("Invoice Category"),
            options:
                "B2B, SEZ, DE\nB2C (Large)\nExports\nB2C (Others)\nNil-Rated, Exempted, Non-GST\nCredit/Debit Notes (Registered)\nCredit/Debit Notes (Unregistered)",
            on_change(report) {
                report.set_filter_value("invoice_sub_category", "");
                set_sub_category_options(report);
            },
            depends_on:
                'eval:doc.summary_by=="Summary by HSN" || doc.summary_by=="Summary by Item"',
        },
        {
            fieldtype: "Autocomplete",
            fieldname: "invoice_sub_category",
            label: __("Invoice Sub Category"),
            depends_on:
                'eval:doc.summary_by=="Summary by HSN" || doc.summary_by=="Summary by Item"',
        },
    ],

    formatter: (value, row, column, data, default_formatter) => {
        value = default_formatter(value, row, column, data);
        if (data && data.indent === 0) {
            let $value = $(`<span>${value}</span>`).css("font-weight", "bold");
            value = $value.wrap("<p></p>").parent().html();
        }

        return value;
    },

    // Override datatable hook for column total calculation
    get_datatable_options(datatable_options) {
        datatable_options.hooks = {
            columnTotal: custom_report_column_total,
        };

        return datatable_options;
    },
};

function set_sub_category_options(report) {
    const invoice_category = frappe.query_report.get_filter_value("invoice_category");
    report
        .get_filter("invoice_sub_category")
        .set_data(INVOICE_TYPE[invoice_category] || []);

    if (invoice_category && INVOICE_TYPE[invoice_category].length === 1) {
        report.set_filter_value(
            "invoice_sub_category",
            INVOICE_TYPE[invoice_category][0]
        );
    }
}

custom_report_column_total = function (...args) {
    const summary_by = frappe.query_report.get_filter_value("summary_by");
    if (summary_by !== "Overview")
        return frappe.utils.report_column_total.apply(this, args);

    const column_field = args[1].column.fieldname;
    if (column_field === "description") return;

    const total = this.datamanager.data.reduce((acc, row) => {
        if (row.indent !== 1) acc += row[column_field] || 0;
        return acc;
    }, 0);

    return total;
};

File: india_compliance/gst_india/report/gst_sales_register_beta/gst_sales_register_beta.py
# Copyright (c) 2024, Resilient Tech and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate

from india_compliance.gst_india.utils.gstr_1.gstr_1_data import GSTR1Invoices


def execute(filters=None):
    filters = validate_filters(filters)
    data = get_data(filters)
    columns = get_columns(filters)

    return columns, data


def validate_filters(filters):
    filters = frappe._dict(filters)
    filters["from_date"] = filters.date_range[0]
    filters["to_date"] = filters.date_range[1]

    if filters.from_date and filters.to_date:
        if getdate(filters.from_date) > getdate(filters.to_date):
            frappe.throw(
                _("From Date must be before To Date"), title=_("Invalid Filter")
            )

    return filters


def get_data(filters):
    _class = GSTR1Invoices(filters)
    invoices = []

    if filters.summary_by == "Summary by Item":
        invoices = _class.get_invoices_for_item_wise_summary()

    if filters.summary_by == "Summary by HSN":
        invoices = _class.get_invoices_for_hsn_wise_summary()

    if filters.summary_by == "Overview":
        return _class.get_overview()

    if filters.invoice_category:
        return _class.get_filtered_invoices(
            invoices, filters.invoice_category, filters.invoice_sub_category
        )

    _class.process_invoices(invoices)

    return invoices


def get_columns(filters):
    columns = []

    if filters.summary_by == "Overview":
        columns.extend(
            [
                {"label": _("Description"), "fieldname": "description", "width": "240"},
                {
                    "label": _("No. of records"),
                    "fieldname": "no_of_records",
                    "width": "120",
                    "fieldtype": "Int",
                },
                {
                    "label": _("Taxable Value"),
                    "fieldname": "taxable_value",
                    "width": "120",
                    "fieldtype": "Currency",
                    "options": "Company:company:default_currency",
                },
                {
                    "label": _("IGST Amount"),
                    "fieldname": "igst_amount",
                    "width": "120",
                    "fieldtype": "Currency",
                    "options": "Company:company:default_currency",
                },
                {
                    "label": _("CGST Amount"),
                    "fieldname": "cgst_amount",
                    "width": "120",
                    "fieldtype": "Currency",
                    "options": "Company:company:default_currency",
                },
                {
                    "label": _("SGST Amount"),
                    "fieldname": "sgst_amount",
                    "width": "120",
                    "fieldtype": "Currency",
                    "options": "Company:company:default_currency",
                },
                {
                    "label": _("Total Cess Amount"),
                    "fieldname": "total_cess_amount",
                    "width": "120",
                    "fieldtype": "Currency",
                    "options": "Company:company:default_currency",
                },
            ]
        )

        return columns

    if not filters.company_gstin:
        columns.append(
            {
                "label": _("Company GSTIN"),
                "fieldname": "company_gstin",
                "width": 180,
            },
        )
    columns.extend(
        [
            {
                "label": _("Posting Date"),
                "fieldname": "posting_date",
                "width": 120,
            },
            {
                "label": _("Invoice Number"),
                "fieldname": "invoice_no",
                "fieldtype": "Link",
                "options": "Sales Invoice",
                "width": 150,
            },
            {
                "label": _("Customer Name"),
                "fieldname": "customer_name",
                "fieldtype": "Link",
                "options": "Customer",
                "width": 150,
            },
            {
                "label": _("GST Category"),
                "fieldname": "gst_category",
                "width": 120,
                "fieldtype": "Data",
            },
            {
                "label": _("Billing Address GSTIN"),
                "fieldname": "billing_address_gstin",
                "width": 180,
                "fieldtype": "Data",
            },
            {
                "label": _("Place of Supply"),
                "fieldname": "place_of_supply",
                "width": 120,
                "fieldtype": "Data",
            },
        ]
    )

    gst_settings = frappe.get_doc("GST Settings")

    if gst_settings.enable_reverse_charge_in_sales:
        columns.append(
            {
                "label": _("Is Reverse Charge"),
                "fieldname": "is_reverse_charge",
                "fieldtype": "Check",
                "width": 120,
            }
        )

    if gst_settings.enable_overseas_transactions:
        columns.append(
            {
                "label": _("Is Export with GST"),
                "fieldname": "is_export_with_gst",
                "fieldtype": "Check",
                "width": 120,
            }
        )

    columns.extend(
        [
            {
                "label": _("Is Return"),
                "fieldname": "is_return",
                "fieldtype": "Check",
                "width": 120,
            },
            {
                "label": _("Is Debit Note"),
                "fieldname": "is_debit_note",
                "fieldtype": "Check",
                "width": 120,
            },
        ]
    )

    if filters.summary_by == "Summary by Item":
        columns.append(
            {
                "label": _("Item Code"),
                "fieldname": "item_code",
                "fieldtype": "Link",
                "options": "Item",
                "width": 180,
            }
        )

    columns.extend(
        [
            {
                "label": _("HSN Code"),
                "fieldname": "gst_hsn_code",
                "fieldtype": "Link",
                "options": "GST HSN Code",
                "width": 120,
            },
            {
                "label": _("UOM"),
                "fieldname": "uom",
                "fieldtype": "Data",
                "width": 100,
            },
            {
                "label": _("GST Treatment"),
                "fieldname": "gst_treatment",
                "width": 120,
                "fieldtype": "Data",
            },
            {
                "label": _("Taxable Value"),
                "fieldname": "taxable_value",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("GST Rate"),
                "fieldname": "gst_rate",
                "width": 120,
                "fieldtype": "Data",
            },
            {
                "label": _("CGST Amount"),
                "fieldname": "cgst_amount",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("SGST Amount"),
                "fieldname": "sgst_amount",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("IGST Amount"),
                "fieldname": "igst_amount",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("Total Cess Amount"),
                "fieldname": "total_cess_amount",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("Total Tax"),
                "fieldname": "total_tax",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("Total Amount"),
                "fieldname": "total_amount",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("Retured Invoice Total"),
                "fieldname": "returned_invoice_total",
                "width": 120,
                "fieldtype": "Currency",
                "options": "Company:company:default_currency",
            },
            {
                "label": _("Invoice Type"),
                "fieldname": "invoice_type",
                "width": 120,
                "fieldtype": "Data",
            },
        ]
    )
    if not filters.invoice_category:
        columns.append(
            {
                "label": _("Invoice Category"),
                "fieldname": "invoice_category",
                "width": 120,
                "fieldtype": "Data",
            }
        )
    if not filters.invoice_sub_category:
        columns.append(
            {
                "label": _("Invoice Sub Category"),
                "fieldname": "invoice_sub_category",
                "width": 120,
                "fieldtype": "Data",
            }
        )

    return columns
